### ISSUE TITLE
Torch Profiler does not count FLOPs for backward pass 

### ISSUE THREAD

**Issue by mariushobbhahn** [pytorch/pytorch#69782]
### 🐛 Describe the bug

I wanted to measure the FLOPs of forward and backward pass with the Pytorch Profiler. However, the backward pass doesn't seem to be tracked. 

```Python
from torch.profiler import profile
import torch
import torch.optim as optim
import torchvision.models as models

# setup model input and target
model = models.resnet18(num_classes=10).cuda()
inputs=torch.randn((1,3,224,224)).cuda() # dummy input
y = torch.randint(10, size=(1,)).cuda()  # dummy target

# warm up cuda memory allocator, recommended here: https://github.com/pytorch/pytorch/blob/master/torch/autograd/profiler.py
outputs = model(inputs)

### forward only
with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
    with_flops=True) as prof:
        outputs = model(inputs)

events = prof.events()
forward_flops = sum([int(evt.flops) for evt in events]) 
print("forward flops: ", forward_flops)
# forward flops:  3627132948

### forward + backward
optimizer = optim.SGD(model.parameters(), lr=0.01)
criterion = torch.nn.CrossEntropyLoss()
# warm up cuda memory allocator
optimizer.zero_grad()
outputs = model(inputs)
loss = criterion(outputs, y)
loss.backward()
optimizer.step()

### measure only backward pass
optimizer.zero_grad()
outputs = model(inputs)
with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
    with_flops=True) as prof:
        outputs = model(inputs) 
        loss = criterion(outputs, y)
        loss.backward()
        optimizer.step()
            
events = prof.events()
fwbw_flops = sum([int(evt.flops) for evt in events]) 
print("forward + backward flops: ", fwbw_flops)
# forward + backward flops:  3628081684
print("forward + backward - forward:", fwbw_flops - forward_flops)
# forward + backward - forward: 948736
```

The FLOPs for the forward pass look plausible for this kind of setup. The backward pass is usually around 2x that of the forward pass. In this case, however, it is multiple orders of magnitude smaller and thus probably an incorrect count. 

Is there any way to count the backward pass as well?

### Versions

I used Google Colab which currently runs on torch version "1.10.0+cu111"

**Comment by sonelu** [comment]
Actually that is very plausible: CNNs have a very high ratio FLOPs/params due to the fact that only the size and the number of kernels define the number of parameters, while FLOPs are also dependent on the input size. The backward pass FLOPs is dependent on the number of parameters, so for CNNs should be low.

You can double check the results by using a different input size (ex. 320x320) and you should see and increase in the reported FLOPS roughly in line with the size change (320x320/224x224) while the number of parameters and therefore the backward FLOPs would stay the same.

**Comment by mariushobbhahn** [comment]
**First of all**, I'm not sure why the backward pass should only depend on the parameters and not on all operations of the forward pass. If we look at the [equations of backpropagation](http://neuralnetworksanddeeplearning.com/chap2.html#:~:text=The%20backpropagation%20equations%20provide%20us,%3D%CF%83(zl).):

![backprop_equations](https://user-images.githubusercontent.com/22917980/145719412-28a94d32-eca0-4130-8f37-8093bf6d96d1.png)

We can see that during backdrop we compute the contribution of every single operation from the forward pass to the loss. Therefore, the backward pass should never have less FLOP than the forward pass. 

**Secondly,** even if you were right and we were bound by the number of parameters, the numbers still don't add up. Resnet18 has about 11 Million trainable parameters. The difference between forward and backward pass as shown above was 948736. This is about 11 times less than the number of parameters. So somehow we do a backward pass but only update less than 10 percent of all weights? That seems wrong. 

I have played around with different networks, batch sizes, etc. and I'm quite convinced that there has to be a bug in the measurement of the backward pass. I just don't know where.

**Comment by ngimel** [comment]
Thanks for reporting, we should improve the coverage for the backward ops. 
The FLOPS for the backward pass even for CNNs should be roughly 2x the flops of the forward pass (unless there are strided or dilated convolutions).

**Comment by WenheLI** [comment]
Hi, just wondering if there will be any way to count the backward flops in pytorch executation?

**Comment by robieta** [comment]
Yes, we are working on improvements to FLOP counting, and improved coverage (including backward ops) is high on the list.

**Comment by jo-jstrm** [comment]
Is there any update on this issue? I would like to compare multiple training strategies by their total used FLOPs, among other metrics. Getting accurate measurements from the profiler would come in quite handy.

**Comment by Markel** [comment]
Same here, just wanting to know if any update happened :)
